File: backend/src/Services/BillService.cs
using API.Entities;
using API.Interfaces.Repositories;
using API.Types;

namespace API.Services;

public class BillService(IBillRepository billRepository) {

    public async Task<Page<Bill>> GetAll(string filter, int page, int limit) {

        Page<Bill> bills = await billRepository.GetAll(filter, page, limit);

        return bills;

    }

    public async Task<Page<Bill>> GetBillsByApartment(int apartmentId, string filter, int page, int limit) {

        Page<Bill> bills = await billRepository.GetByApartmentId(apartmentId, filter, page, limit);

        return bills;

    }

    public async Task<Bill?> GetBillById(int id) {

        Bill? bill = await billRepository.GetById(id);

        return bill;

    }

    public async Task<Bill?> CreateBill(int billTypeId, int apartmentId, int month, string fileName, string filePath) {

        Bill bill = new() {
            BillTypeId = billTypeId,
            ApartmentId = apartmentId,
            Month = month,
            FileName = fileName,
            FilePath = filePath
        };

        try {
            return await billRepository.Create(bill);
        } catch {
            return null;
        }

    }

    public async Task<Bill?> UpdateBill(int id, int? billTypeId, int? apartmentId, int? month, string? fileName, string? filePath) {

        Bill? bill = await billRepository.GetById(id);

        if(bill == null) {
            return null;
        }

        bill.BillTypeId = billTypeId ?? bill.BillTypeId;
        bill.ApartmentId = apartmentId ?? bill.ApartmentId;
        bill.Month = month ?? bill.Month;
        bill.FileName = fileName ?? bill.FileName;
        bill.FilePath = filePath ?? bill.FilePath;

        try {
            return await billRepository.Update(bill);
        } catch {
            return null;
        }

    }

    public async Task<Bill?> DeleteBill(int id) {

        Bill? bill = await billRepository.GetById(id);

        if(bill == null) {
            return null;
        }

        try {
            return await billRepository.Delete(bill);
        } catch {
            return null;
        }

    }

}

File: frontend/src/services/bill.service.ts
import { APIError } from "@/models/api-error.models";
import { Bill } from "@/models/bill.model";
import { Page } from "@/models/page";
import { Service } from "@/services/service";
import { optional } from "@/types/optional";

export class BillService extends Service {

    static async getBillsByApartment(apartmentId : number, filter : string, page : number, limit : number) {

        const response = await this.axios.get<Page<Bill> | APIError>(`/api/bill/apartment/${apartmentId}`, {
            params: {
                filter,
                page,
                limit
            }
        });

        return response.data;

    }

    static async createBill(billTypeId : number, apartmentId : number, month : number, year : number, file : File) {

        const formData = new FormData();

        formData.append("billTypeId", billTypeId.toString());
        formData.append("apartmentId", apartmentId.toString());
        formData.append("month", month.toString());
        formData.append("year", year.toString());

        if(file) {
            formData.append("file", file);
        }

        const response = await this.axios.post<Bill | APIError>("/api/bill", formData);

        return response.data;

    }

    static async updateBill(id : number, billTypeId : optional<number>, apartmentId : optional<number>, month : optional<number>, year : optional<number>, file : optional<File>) {

        const formData = new FormData();

        if(billTypeId) {
            formData.append("billTypeId", billTypeId.toString());
        }

        if(apartmentId) {
            formData.append("apartmentId", apartmentId.toString());
        }

        if(month) {
            formData.append("month", month.toString());
        }

        if(year) {
            formData.append("year", year.toString());
        }

        if(file) {
            formData.append("file", file);
        }

        const response = await this.axios.put<Bill | APIError>(`/api/bill/${id}`, formData);

        return response.data;

    }

    static async deleteBill(id : number) {

        const response = await this.axios.delete<Bill | APIError>(`/api/bill/${id}`);

        return response.data;

    }

}